### PR TITLE
Jetpack Sync: Add sync action for network site enable and disable themes

### PR DESCRIPTION
**Pull request by gititon**
Jetpack Sync: Add sync action for network/multisite enable and disable themes

#### Changes proposed in this Pull Request:

Add sync action for network/multisite enable and disable themes

#### Testing instructions:

phpunit runs new tests when on multisite

#### Proposed changelog entry for your changes:
Addition of jetpack_network_disabled_themes and jetpack_network_enabled_themes hooks

### CODE REVIEW

File: sync/class.jetpack-sync-module-themes.php [hunk 1]
```diff
@@ -25,6 +28,61 @@ public function init_listeners( $callable ) {
...
+			}
+
+			/**
+			 * Trigger action to alert $callable sync listener that network themes were disabled
+			 *
+			 * @since 5.0.0
+			 *
+			 * @param mixed $disabled_themes, Array of info about network disabled themes
+			 */
+			do_action( 'jetpack_network_disabled_themes', $disabled_themes );
```
Comment (lezama) [inline sync/class.jetpack-sync-module-themes.php]: we could send the enabled ones too? what do you think?

File: sync/class.jetpack-sync-module-themes.php [hunk 2]
```diff
@@ -25,6 +28,61 @@ public function init_listeners( $callable ) {
...
+			return;
+		}
+
+		if ( 'enable-selected' === $action || 'enable' === $action ) {
+			$enabled_theme_names = array_keys( array_diff_key( $value, $old_value ) );
+			$enabled_themes = array();
+			foreach ( $enabled_theme_names as $name ) {
```
Comment (lezama) [inline sync/class.jetpack-sync-module-themes.php]: we could probably extract this to its own method

File: sync/class.jetpack-sync-module-themes.php (hunk 1 above)
Comment (gititon) [inline sync/class.jetpack-sync-module-themes.php]: Themes can't be both changed from enabled to disabled and changed from disabled to enabled by the same user action. When themes are disabled, the Activity Log needs to know which themes were disabled. When themes are enabled, the Activity Log needs to know which themes were enabled. Why would we send themes that are unrelated to the activity/action?

Comment (lezama) [inline sync/class.jetpack-sync-module-themes.php]: not sure really, I thought it could be useful to always know the active ones

File: sync/class.jetpack-sync-module-themes.php [hunk 3]
```diff
@@ -25,6 +28,46 @@ public function init_listeners( $callable ) {
...
+	public function sync_network_allowed_themes_change( $option, $value, $old_value, $network_id ) {
+		if ( count( $old_value ) > count( $value ) )  {
+			$disabled_theme_names = array_keys( array_diff_key( $old_value, $value ) );
+			$disabled_themes = $this->get_multisite_changed_themes( $disabled_theme_names );
+			/**
+			 * Trigger action to alert $callable sync listener that network themes were disabled
+			 *
+			 * @since 5.0.0
+			 *
+			 * @param mixed $disabled_themes, Array of info about network disabled themes
```
Comment (lezama) [inline sync/class.jetpack-sync-module-themes.php]: we should add the second parameter too

File: sync/class.jetpack-sync-module-themes.php [hunk 4]
```diff
@@ -25,6 +28,46 @@ public function init_listeners( $callable ) {
...
+		}
+
+		$enabled_theme_names = array_keys( array_diff_key( $value, $old_value ) );
+		$enabled_themes = $this->get_multisite_changed_themes( $enabled_theme_names );
+		/**
+		 * Trigger action to alert $callable sync listener that network themes were enabled
+		 *
+		 * @since 5.0.0
+		 *
+		 * @param mixed $enabled_themes , Array of info about network enabled themes
```
Comment (lezama) [inline sync/class.jetpack-sync-module-themes.php]: we should add the second parameter too

File: sync/class.jetpack-sync-module-themes.php [hunk 5]
```diff
@@ -25,6 +28,46 @@ public function init_listeners( $callable ) {
...
+		 */
+		do_action( 'jetpack_network_enabled_themes', $enabled_themes, array_keys( $value ) );
+	}
+
+	private function get_multisite_changed_themes( $theme_names ) {
```
Comment (lezama) [inline sync/class.jetpack-sync-module-themes.php]: we could call this method `get_theme_details`, we could use it with non multisite themes too

File: tests/php/sync/test_class.jetpack-sync-themes.php [hunk 6]
```diff
@@ -123,6 +123,126 @@ public function test_changed_theme_is_synced() {
...
+		$theme_slugs = array_keys( $themes );
+
+		//Test enable single theme
+
+		/**
+		 * This filter is already documented in wp-includes/option.php
+		 *
+		 * Note that 'allowedthemes' is dynamic, i.e. do_action is called on "update_site_option_{$option}"
+		 *
+		 * @since 2.9.0
```
Comment (enejb) [inline tests/php/sync/test_class.jetpack-sync-themes.php]: I would remove the since part. Since the version number refers to WP and not Jetpack.

File: tests/php/sync/test_class.jetpack-sync-themes.php [hunk 7]
```diff
@@ -123,6 +123,126 @@ public function test_changed_theme_is_synced() {
...
+		$this->server_event_storage->reset();
+
+		//Test disable single theme
+
+		/**
+		 * This filter is already documented in wp-includes/option.php
+		 *
+		 * Note that 'allowedthemes' is dynamic, i.e. do_action is called on "update_site_option_{$option}"
+		 *
+		 * @since 2.9.0
```
Comment (enejb) [inline tests/php/sync/test_class.jetpack-sync-themes.php]: same as here

File: tests/php/sync/test_class.jetpack-sync-themes.php [hunk 8]
```diff
@@ -123,6 +123,126 @@ public function test_changed_theme_is_synced() {
...
+		$this->server_event_storage->reset();
+
+		//Test disable multiple themes
+
+		/**
+		 * This filter is already documented in wp-includes/option.php
+		 *
+		 * Note that 'allowedthemes' is dynamic, i.e. do_action is called on "update_site_option_{$option}"
+		 *
+		 * @since 2.9.0
```
Comment (enejb) [inline tests/php/sync/test_class.jetpack-sync-themes.php]: here also

File: tests/php/sync/test_class.jetpack-sync-themes.php [hunk 9]
```diff
@@ -123,6 +123,126 @@ public function test_changed_theme_is_synced() {
...
+		$theme_slugs = array_keys( $themes );
+
+		//Test enable multiple themes
+
+		/**
+		 * This filter is already documented in wp-includes/option.php
+		 *
+		 * Note that 'allowedthemes' is dynamic, i.e. do_action is called on "update_site_option_{$option}"
+		 *
+		 * @since 2.9.0
```
Comment (enejb) [inline tests/php/sync/test_class.jetpack-sync-themes.php]: here too

File: sync/class.jetpack-sync-module-themes.php [hunk 10]
```diff
@@ -25,6 +28,48 @@ public function init_listeners( $callable ) {
...
+			$disabled_themes = $this->get_theme_details_for_slugs( $disabled_theme_names );
+			/**
+			 * Trigger action to alert $callable sync listener that network themes were disabled
+			 *
+			 * @since 5.0.0
+			 *
+			 * @param mixed $disabled_themes, Array of info about network disabled themes
+			 * @param mixed array_keys( $value ), Array of slugs of all enabled themes
+			 */
+			do_action( 'jetpack_network_disabled_themes', $disabled_themes, array_keys( $value ) );
```
Comment (enejb) [inline sync/class.jetpack-sync-module-themes.php]: Can we make array_keys( $value ) more descriptive. Something like 
`$enabled_theme_slugs` would read better

File: sync/class.jetpack-sync-module-themes.php [hunk 11]
```diff
@@ -25,6 +28,48 @@ public function init_listeners( $callable ) {
...
+		$enabled_themes = $this->get_theme_details_for_slugs( $enabled_theme_names );
+		/**
+		 * Trigger action to alert $callable sync listener that network themes were enabled
+		 *
+		 * @since 5.0.0
+		 *
+		 * @param mixed $enabled_themes , Array of info about network enabled themes
+		 * @param mixed array_keys( $value ), Array of slugs of all enabled themes
+		 */
+		do_action( 'jetpack_network_enabled_themes', $enabled_themes, array_keys( $value ) );
```
Comment (enejb) [inline sync/class.jetpack-sync-module-themes.php]: Can we make array_keys( $value ) more descriptive. Something like 
`$enabled_theme_slugs` would read better

File: sync/class.jetpack-sync-module-themes.php [hunk 12]
```diff
@@ -25,6 +28,48 @@ public function init_listeners( $callable ) {
 		add_action( 'jetpack_widget_reordered', $callable );
 	}
 
+	public function sync_network_allowed_themes_change( $option, $value, $old_value, $network_id ) {
+		if ( count( $old_value ) > count( $value ) )  {
+			$disabled_theme_names = array_keys( array_diff_key( $old_value, $value ) );
+			$disabled_themes = $this->get_theme_details_for_slugs( $disabled_theme_names );
```
Comment (enejb) [inline sync/class.jetpack-sync-module-themes.php]: Something like $newly_disabled_themes might be a bit more descriptive.

File: tests/php/sync/test_class.jetpack-sync-themes.php [hunk 13]
```diff
@@ -123,6 +123,126 @@ public function test_changed_theme_is_synced() {
...
+		 */
+		do_action( 'update_site_option_allowedthemes', 'allowedthemes', $themes, array(), 0 );
+		$this->sender->do_sync();
+
+		$event_data = $this->server_event_storage->get_most_recent_event( 'jetpack_network_enabled_themes' );
+
+		foreach ( $test_themes as $theme ) {
```
Comment (enejb) [inline tests/php/sync/test_class.jetpack-sync-themes.php]: Lets also test that the slug is also being set.

File: tests/php/sync/test_class.jetpack-sync-themes.php [hunk 14]
```diff
@@ -123,6 +123,126 @@ public function test_changed_theme_is_synced() {
...
+		 * Note that 'allowedthemes' is dynamic, i.e. do_action is called on "update_site_option_{$option}"
+		 *
+		 * @since 2.9.0
+		 */
+		do_action( 'update_site_option_allowedthemes', 'allowedthemes', array(), $themes, 0 );
+		$this->sender->do_sync();
+
+		$event_data = $this->server_event_storage->get_most_recent_event( 'jetpack_network_disabled_themes' );
+
+		foreach ( $test_themes as $theme ) {
```
Comment (enejb) [inline tests/php/sync/test_class.jetpack-sync-themes.php]: lets test for the slug here too.

File: sync/class.jetpack-sync-module-themes.php (hunk 3 above)
Comment (gititon) [inline sync/class.jetpack-sync-module-themes.php]: Resolved

File: sync/class.jetpack-sync-module-themes.php (hunk 4 above)
Comment (gititon) [inline sync/class.jetpack-sync-module-themes.php]: Resolved

File: tests/php/sync/test_class.jetpack-sync-themes.php (hunk 6 above)
Comment (gititon) [inline tests/php/sync/test_class.jetpack-sync-themes.php]: Done

File: tests/php/sync/test_class.jetpack-sync-themes.php (hunk 7 above)
Comment (gititon) [inline tests/php/sync/test_class.jetpack-sync-themes.php]: Done

File: tests/php/sync/test_class.jetpack-sync-themes.php (hunk 8 above)
Comment (gititon) [inline tests/php/sync/test_class.jetpack-sync-themes.php]: Done

File: tests/php/sync/test_class.jetpack-sync-themes.php (hunk 9 above)
Comment (gititon) [inline tests/php/sync/test_class.jetpack-sync-themes.php]: Done

File: sync/class.jetpack-sync-module-themes.php (hunk 10 above)
Comment (gititon) [inline sync/class.jetpack-sync-module-themes.php]: Done

File: sync/class.jetpack-sync-module-themes.php (hunk 11 above)
Comment (gititon) [inline sync/class.jetpack-sync-module-themes.php]: Done

File: sync/class.jetpack-sync-module-themes.php (hunk 12 above)
Comment (gititon) [inline sync/class.jetpack-sync-module-themes.php]: done

File: tests/php/sync/test_class.jetpack-sync-themes.php (hunk 13 above)
Comment (gititon) [inline tests/php/sync/test_class.jetpack-sync-themes.php]: done

File: tests/php/sync/test_class.jetpack-sync-themes.php (hunk 14 above)
Comment (gititon) [inline tests/php/sync/test_class.jetpack-sync-themes.php]: done
